NN: Integrate CMSIS-NN 1.3.0 (CMSIS 5.7.0)

trembel opened the pull request:
This commit integrates the newly added CMSIS-NN 1.3.0 (part of CMSIS
5.7.0 release) component to the Zephyr build system.

It refers to https://github.com/zephyrproject-rtos/cmsis/pull/6.
Changes were made to `Kconfig.cmsis` and `west.yml`,
`Kconfig.cmsis_nn` is added.

Signed-off-by: Silvano Cortesi <cortesis@student.ethz.ch>

stephanosio commented on modules/Kconfig.cmsis:
```diff
@@ -28,3 +28,13 @@ menuconfig CMSIS_DSP
+menuconfig CMSIS_NN
+	bool "CMSIS-NN Library Support"
+	depends on CPU_CORTEX
+	depends on NEWLIB_LIBC
```
Does CMSIS-NN directly depend on the functions that are not available in the Zephyr minimal libc?

If not, `depends on NEWLIB_LIBC` should be removed (note that `select CMSIS_DSP`, which does depend on the newlib, will resolve the dependency anyway).

stephanosio commented on modules/Kconfig.cmsis:
```diff
@@ -28,3 +28,13 @@ menuconfig CMSIS_DSP
+	bool "CMSIS-NN Library Support"
+	depends on CPU_CORTEX
+	depends on NEWLIB_LIBC
+	select CMSIS_DSP
```
Note that simply selecting `CMSIS_DSP` symbol may not necessarily select the CMSIS-DSP components that CMSIS-NN depends on. This should resolve the exact CMSIS-DSP component dependec{y,ies}.

trembel replied:
Haven't I done that already starting from line line 37?

stephanosio replied:
I was looking at the first commit, which did not have these.

stephanosio commented on modules/Kconfig.cmsis:
```diff
@@ -34,7 +34,18 @@ menuconfig CMSIS_NN
 	depends on NEWLIB_LIBC
 	select CMSIS_DSP
-
+	select CMSIS_DSP_BASICMATH
+	select CMSIS_DSP_COMPLEXMATH
+	select CMSIS_DSP_CONTROLLER
+	select CMSIS_DSP_FASTMATH
+	select CMSIS_DSP_FILTERING
+	select CMSIS_DSP_MATRIX
+	select CMSIS_DSP_STATISTICS
+	select CMSIS_DSP_SUPPORT
+	select CMSIS_DSP_TRANSFORM
+	select CMSIS_DSP_SVM
+	select CMSIS_DSP_BAYES
+	select CMSIS_DSP_DISTANCE
```
Pretty sure not all these are required.

trembel replied:
Do you know of an easy way to find out which ones are used?

stephanosio replied:
I am afraid the "trial-and-error approach" would be the quickest/easiest in this case.

trembel replied:
I've made `grep -rn './' -e 'function_name'` for all functions listened on https://arm-software.github.io/CMSIS_5/DSP/html/modules.html in the CMSIS/NN folder.

None of all them are used - can this be correct or am I missing something?

trembel replied:
After further investigation I assume, that they only need the Header `arm_math.h` for type definitions. I'll therefore set all CMSIS_DSP components, except CMSIS_DSP itself, to n

stephanosio replied:
> After further investigation I assume, that they only need the Header `arm_math.h` for type definitions. I'll therefore set all CMSIS_DSP components, except CMSIS_DSP itself, to n

I did a quick scan of the CMSIS-NN source code and that does seem to be the case. We will see what kind of errors we get (if any) when we build the CMSIS-NN tests in the future.

stephanosio commented on modules/Kconfig.cmsis_nn:
```diff
@@ -0,0 +1,77 @@
+comment "Components"
+
+config CMSIS_NN_ACTIVATION
+       bool "Activation"
+       default y
+       help
```
Please fix indentation. These should be indented with one TAB character.

stephanosio commented on modules/Kconfig.cmsis_nn:
```diff
@@ -0,0 +1,77 @@
+       bool "Activation"
+       default y
+       help
+	This option enables the NN libraries for the activation layers
```
Please fix indentation. This should be indented with one TAB + two SPACEs.

stephanosio commented on modules/Kconfig.cmsis_nn:
```diff
@@ -0,0 +1,77 @@
+       default y
+       help
+	When off, its default behavior is all tables are included
+
+config CMSIS_NN_POOLING
+       bool "Pooling"
+       default y
+       help
+	This option enables the NN libraries for the pooling layers
+
+config CMSIS_NN_RESHAPE
+       bool "Reshape"
+       default y
+       help
+	This option enables the NN libraries for the reshape layers
+
+config CMSIS_NN_SOFTMAX
+       bool "Softmax"
+       default y
+       help
+	This option enables the NN libraries for the softmax layers
+
+# comment "Instruction Set"
+# NOTE: These configurations should eventually be derived from the arch ISA and
+#       FP support configurations.
+
+# config CMSIS_DSP_MVEI
+#	bool "MVE-I Instructions"
+#	depends on CPU_CORTEX_M
+#	help
+#	  This option enables the ARMv8.1-M M-Profile Vector Extension (MVE)
+#	  integer instruction set.
+
+comment "Features"
+
+config CMSIS_DSP_AUTOVECTORIZE
+	bool "Auto Vectorize"
+	help
+	  This option prefers autovectorizable code to one using C intrinsics
+	  in the DSP functions.
```
Indentation issue for all these.

stephanosio replied:
These should not be here. They are already defined by CMSIS-DSP.

stephanosio reviewed (changes requested):
Please fix the commits:
* Commit title and message format
* There should be no merge commit in a PR
* Please amend the incorrect commits instead of adding new commits to fix the errors in the previous commits

Please read https://docs.zephyrproject.org/latest/contribute/index.html.

stephanosio reviewed (changes requested):
There should be one or more test(s) and/or sample(s) to ensure that this feature stays in working shape.

trembel commented:
> Please fix the commits:
> 
>     * Commit title and message format
> 
>     * There should be no merge commit in a PR
> 
>     * Please amend the incorrect commits instead of adding new commits to fix the errors in the previous commits
> 
> 
> Please read https://docs.zephyrproject.org/latest/contribute/index.html.



> Please fix the commits:
> 
>     * Commit title and message format
> 
>     * There should be no merge commit in a PR
> 
>     * Please amend the incorrect commits instead of adding new commits to fix the errors in the previous commits
> 
> 
> Please read https://docs.zephyrproject.org/latest/contribute/index.html.

How can I remove the merge? By making rebase and delete all the commits coming from the merge?

trembel commented:
> There should be one or more test(s) and/or sample(s) to ensure that this feature stays in working shape.

I've at the moment not enough time (for the next two months) to add tests, so I'll leave this PR open.

stephanosio commented:
> > There should be one or more test(s) and/or sample(s) to ensure that this feature stays in working shape.
> 
> I've at the moment not enough time (for the next two months) to add tests, so I'll leave this PR open.

@trembel Alright, I will see if I can find some time to add the tests for CMSIS-NN; after all, it uses the same test framework as the CMSIS-DSP and it should be easy enough for me to write tests for them.

stephanosio reviewed (commented):
Looking much better now.

Can you update the commit title and message to include the CMSIS-NN version number here as well?

```
modules: cmsis: Integrate CMSIS-NN 1.3.0 (CMSIS 5.7.0)

This commit integrates the newly added CMSIS-NN 1.3.0 (part of CMSIS
5.7.0 release) component to the Zephyr build system.
...
```

Commit:
modules: cmsis: Integrate CMSIS-NN 1.3.0 (CMSIS 5.7.0)

This commit integrates the newly added CMSIS-NN 1.3.0 (part of CMSIS
5.7.0 release) component to the Zephyr build system.

It refers to https://github.com/zephyrproject-rtos/cmsis/pull/6.
Changes were made to `Kconfig.cmsis` and `west.yml`,
`Kconfig.cmsis_nn` is added.

Signed-off-by: Silvano Cortesi <cortesis@student.ethz.ch>

stephanosio reviewed (commented):
In the commit title, the project-wide convention is to use `:` instead of `-`:
i.e. `modules: cmsis: Integrate CMSIS-NN 1.3.0 (CMSIS 5.7.0)`

Could you fix this?

Other than that, it looks ok to me. Just need some tests to validate that it works.

p.s. the help documentation aspect of the Kconfig symbols could probably use some improvements, but I won't block this PR for that.

trembel commented:
> In the commit title, the project-wide convention is to use `:` instead of `-`:
> i.e. `modules: cmsis: Integrate CMSIS-NN 1.3.0 (CMSIS 5.7.0)`
> 
> Could you fix this?
> 
> Other than that, it looks ok to me. Just need some tests to validate that it works.
> 
> p.s. the help documentation aspect of the Kconfig symbols could probably use some improvements, but I won't block this PR for that.

Fixed this and improved some of the help messages for Kconfig options.

galak commented:
We do need at least one sample/test that uses this.

trembel commented:
> We do need at least one sample/test that uses this.

I'll probably dive into it next month.

microbuilder commented on modules/Kconfig.cmsis_nn:
```diff
@@ -0,0 +1,64 @@
+	  It can perform activation layers, including ReLU (Rectified
+	  Linear Unit), sigmoid and tanh.
+
+config CMSIS_NN_BASICMATHS
```
Why `BASICMATHS` with the `S`? I understand the linguistic preference for an `s` in some contexts, but I was testing this out, trying to put a simple test function together for sanitycheck, and the extra `S` caught me off guard since the CMSIS-NN library uses `basicmath` for the folder, doxygen sections, etc. For example: https://github.com/zephyrproject-rtos/cmsis/pull/6/files#diff-2ce6a9fd5ca36dd23269b984185b588dR40

I think consistency is important here and would suggest `CMSIS_NN_BASICMATH`.

trembel replied:
This one is a typo, changed that.

microbuilder commented:
> > We do need at least one sample/test that uses this.
> 
> I'll probably dive into it next month.

If it helps, here is a patch file that adds a simple test, which can be extended to cover other parts of CMSIS-NN.

You can run this via `$ zephyr/scripts/sanitycheck -p mps2_an521 -T zephyr/tests/lib/cmsis_nn/nnsupport/`:

Patch file: [cmsis-nn.txt](https://github.com/zephyrproject-rtos/zephyr/files/4878358/cmsis-nn.txt)

Running this requires making these changes in your other PR: 
- https://github.com/zephyrproject-rtos/cmsis/pull/6#pullrequestreview-442930085
- https://github.com/zephyrproject-rtos/cmsis/pull/6#pullrequestreview-442957686

trembel commented:
> > > We do need at least one sample/test that uses this.
> > 
> > 
> > I'll probably dive into it next month.
> 
> If it helps, here is a patch file that adds a simple test, which can be extended to cover other parts of CMSIS-NN.
> 
> You can run this via `$ zephyr/scripts/sanitycheck -p mps2_an521 -T zephyr/tests/lib/cmsis_nn/nnsupport/`:
> 
> Patch file: [cmsis-nn.txt](https://github.com/zephyrproject-rtos/zephyr/files/4878358/cmsis-nn.txt)
> 
> Running this requires making these changes in your other PR:
> 
>     * [zephyrproject-rtos/cmsis#6 (review)](https://github.com/zephyrproject-rtos/cmsis/pull/6#pullrequestreview-442930085)
> 
>     * [zephyrproject-rtos/cmsis#6 (review)](https://github.com/zephyrproject-rtos/cmsis/pull/6#pullrequestreview-442957686)

That helps me, I'll use that for creating a test file.

carlescufi commented:
@stephanosio @microbuilder could you take another look? https://github.com/zephyrproject-rtos/cmsis/pull/6 is ready for merging

stephanosio reviewed (commented):
Just a reminder/note that this feature still needs some tests.
